File: test/helpers/span_tree_builder.erl
-module(span_tree_builder).

%% API
-export([randomize_span_pattern/1,
         generate_linked_span_trees/2,
         generate_span_tree/2]).

%% otel_tracer.hrl defines tracing marcos (e.g. ?with_span())
-include_lib("opentelemetry_api/include/otel_tracer.hrl").
%% opentelemetry.hrl contains #span_ctx{} record declaration
-include_lib("opentelemetry_api/include/opentelemetry.hrl").

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% API
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


randomize_span_pattern(#{
                         parent_span_id := ParentSpanId,
                         links := Links,
                         events := Events,
                         attributes := Attributes,
                         status := Status
                        } = SpanPattern) ->
    %% randomize the Span pattern a bit
    MaybeRemoveKeys = [kind, status, attributes, events, links],
    RandomKeys = pick_random_items(MaybeRemoveKeys, length(MaybeRemoveKeys), true),
    RemoveKeys = case ParentSpanId of
                     undefined ->
                         %% keep trace_id and span_id for the root span
                         [parent_span_id | RandomKeys];
                     _ ->
                         [trace_id, span_id, parent_span_id | RandomKeys]
                 end,
    maps:without(RemoveKeys,
                 SpanPattern#{
                   status => pick_random_map_keys(Status),
                   attributes => pick_random_map_keys(Attributes),
                   events => pick_random_items(Events, 2, false),
                   links => pick_random_items(Links, 2, false),
                   start_time => '_',
                   end_time => '_'
                  }).


generate_linked_span_trees(SpanTrees, ConvertPatternFn) ->
    FoldFn =
        fun(SpanTree, {SpanTreePatterns, Links}) ->
                {NewSpanTreePattern, NewLinks} =
                    generate_span_tree(SpanTree, Links, ConvertPatternFn),
                {[NewSpanTreePattern | SpanTreePatterns],
                 lists:append(NewLinks, Links)}
        end,
    {SpanTreePatterns, _Links} = lists:foldl(FoldFn, {[], []}, SpanTrees),
    SpanTreePatterns.


generate_span_tree(SpanTree, ConvertPatternFn) ->
    {SpanTreePattern, _Links} = generate_span_tree(SpanTree, [], ConvertPatternFn),
    SpanTreePattern.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% local functions
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


generate_span_tree({#{name := Name} = Span, Children}, Links, ConvertPatternFn) ->
    assert_span_input_data(Span),
    ParentSpanId = get_current_span_id(),
    Kind = maps:get(kind, Span, ?SPAN_KIND_INTERNAL),
    Status = maps:get(status, Span, undefined),
    Attributes = maps:get(attributes, Span, #{}),
    Events = maps:get(events, Span, []),
    SpanLinks = pick_random_items(Links, 4, true),
    ?with_span(
      Name,
      #{kind => Kind, attributes => Attributes, links => SpanLinks},
      fun(SpanCtx) ->
              BranchesAndNewLinks = [ generate_span_tree(Child, Links, ConvertPatternFn)
                                      || Child <- Children ],
              Branches = [ Branch || {Branch, _} <- BranchesAndNewLinks ],
              NewLinks = lists:append([ Link || {_, Link} <- BranchesAndNewLinks ]),
              SpanId = SpanCtx#span_ctx.span_id,
              TraceId = SpanCtx#span_ctx.trace_id,

              %% the links are added to the span in reverse order, but we want to keep
              %% the order of the links in the pattern the same as in the span data.
              SpanLinksPattern =
                  [ link_to_pattern(L) || L <- lists:reverse(SpanLinks) ],

              %% the events are added to the span in reverse order, but we want to keep
              %% the order of the events in the pattern the same as in the span data.
              ?add_events(lists:reverse(Events)),

              set_status(Status),
              CompleteSpanPattern = Span#{
                                      span_id => SpanId,
                                      trace_id => TraceId,
                                      parent_span_id => ParentSpanId,
                                      links => SpanLinksPattern,
                                      events => Events,
                                      attributes => Attributes
                                     },
              SpanPattern = ConvertPatternFn(CompleteSpanPattern),
              NewLink = create_new_link(SpanCtx, Attributes),
              {{SpanPattern, Branches}, [NewLink | NewLinks]}
      end).


assert_span_input_data(Span) ->
    Keys = maps:keys(Span),
    SupportedKeys = [name, kind, status, attributes, events],
    case Keys -- SupportedKeys of
        [] -> ok;
        UnsupportedKeys -> error(span_input_data_keys, UnsupportedKeys)
    end.


set_status(undefined) ->
    ok;
set_status(#{code := Code, message := Message}) ->
    ?set_status(Code, Message).


create_new_link(SpanCtx, Attributes) ->
    RandomAttributes = pick_random_map_keys(Attributes),
    opentelemetry:link(SpanCtx, RandomAttributes).


get_current_span_id() ->
    case ?current_span_ctx of
        undefined -> undefined;
        #span_ctx{span_id = SpanId} -> SpanId
    end.


link_to_pattern(Link) ->
    %% note that the opentelemetry:link() type returned by opentelemetry:link/2 is a map.
    %% While the #span{} record stores links as #link{} records (see record definitions
    %% in the otel_span.hrl header and the definition of the otel_events:t() type), after
    %% converting the #span{} record using the opentelemetry_testing:convert/1 function,
    %% the format of the links in the resulting map is similar to the opentelemetry:link()
    %% type (in fact, it is identical).
    maps:with([trace_id, span_id, attributes], Link).


pick_random_map_keys(Map) when is_map(Map) ->
    KVList = maps:to_list(Map),
    RandomKVList = pick_random_items(KVList, 4, true),
    maps:from_list(RandomKVList);
pick_random_map_keys(Term) ->
    %% covers 'undefined' span status case
    Term.


pick_random_items([], _MaxN, _AllowEmptyResult) ->
    [];
pick_random_items(_, 0, _AllowEmptyResult) ->
    [];
pick_random_items(List, MaxN, AllowEmptyResult) ->
    Length = length(List),
    N = case AllowEmptyResult of
            true -> rand:uniform(min(Length, MaxN) + 1) - 1;
            false -> rand:uniform(min(Length, MaxN))
        end,
    %% in the original span pattern of the randomize_span_pattern/1 function,
    %% the lists of link and event patterns are ordered in the same way as
    %% the data instances (see the generate_span_tree/3 function), and we don't
    %% want a pattern list like this:
    %%    [#{ attributes => #{ attr1 => <<"value1">>}, name => event1 },
    %%     #{ attributes => #{}, name => event2 },
    %%     #{ attributes => #{}, name => event1 }]
    %% to turn into:
    %%    [#{ attributes => #{}, name => event1 },
    %%     #{ attributes => #{ attr1 => <<"value1">>}, name => event1 }]
    %%
    %% let's ensure the random indices are sorted in ascending order, so
    %% the randomize_span_pattern/1 function does not shuffle the patterns
    %% in the event/link list, and no accidental match can happen for a
    %% less restrictive pattern.
    RandomIndexes = lists:usort([ rand:uniform(Length) || _ <- sequence(N) ]),
    [ lists:nth(Index, List) || Index <- RandomIndexes ].


sequence(N) ->
    %% note that lists:seq(1,0) returns []
    lists:seq(1, N).

File: test/span_tree_builder.ex
defmodule SpanTreeBuilder do
  require OpenTelemetry.Tracer, as: Tracer

  #########################################################################
  ## API
  #########################################################################

  def randomize_span_pattern(
        %{
          parent_span_id: parent_span_id,
          links: links,
          events: events,
          attributes: attributes,
          status: status
        } = span_pattern
      ) do
    ## randomize the span pattern a bit
    maybe_remove_keys = [:kind, :status, :attributes, :events, :links]
    random_keys = pick_random_items(maybe_remove_keys, length(maybe_remove_keys), true)

    remove_keys =
      case parent_span_id do
        :undefined ->
          ## keep trace_id and span_id for the root span
          [:parent_span_id | random_keys]

        _ ->
          [:trace_id, :span_id, :parent_span_id | random_keys]
      end

    %{
      span_pattern
      | status: pick_random_map_keys(status),
        attributes: pick_random_map_keys(attributes),
        events: pick_random_items(events, 2, false),
        links: pick_random_items(links, 2, false)
    }
    |> Map.merge(%{start_time: :_, end_time: :_})
    |> Map.drop(remove_keys)
  end

  def generate_linked_span_trees(span_trees, convert_pattern_fn) do
    fold_fn = fn span_tree, {span_tree_patterns, links} ->
      {new_span_tree_pattern, new_links} =
        generate_span_tree(span_tree, links, convert_pattern_fn)

      {[new_span_tree_pattern | span_tree_patterns], Enum.concat(new_links, links)}
    end

    {span_tree_patterns, _links} = List.foldl(span_trees, {[], []}, fold_fn)
    span_tree_patterns
  end

  def generate_span_tree(span_tree, convert_pattern_fn) do
    {span_tree_patterns, _links} = generate_span_tree(span_tree, [], convert_pattern_fn)
    span_tree_patterns
  end

  #########################################################################
  ## local interfaces
  #########################################################################

  defp generate_span_tree({%{name: name} = span, children}, links, convert_pattern_fn) do
    assert_span_input_data(span)
    parent_span_id = get_current_span_id()
    kind = Map.get(span, :kind, :internal)
    status = Map.get(span, :status, :undefined)
    attributes = Map.get(span, :attributes, %{})
    events = Map.get(span, :events, [])
    span_links = pick_random_items(links, 4, true)

    Tracer.with_span name, %{kind: kind, attributes: attributes, links: span_links} do
      {trace_id, span_id} = OpentelemetryTesting.get_span_ids()

      branches_and_new_links =
        Enum.map(children, &generate_span_tree(&1, links, convert_pattern_fn))

      {branches, new_links} = Enum.unzip(branches_and_new_links)

      ## the links are added to the span in reverse order, but we want to keep
      ## the order of the links in the pattern the same as in the span data.
      span_links_pattern = Enum.map(span_links, &link_to_pattern(&1)) |> Enum.reverse()

      ## the events are added to the span in reverse order, but we want to keep
      ## the order of the events in the pattern the same as in the span data.
      Tracer.add_events(Enum.reverse(events))
      set_status(status)

      span_pattern =
        span
        |> Map.merge(%{
          span_id: span_id,
          trace_id: trace_id,
          parent_span_id: parent_span_id,
          attributes: attributes,
          events: events,
          links: span_links_pattern
        })
        |> convert_pattern_fn.()

      link = create_new_link(attributes)
      {{span_pattern, branches}, [link | Enum.concat(new_links)]}
    end
  end

  defp assert_span_input_data(span) do
    supported_keys = [:name, :kind, :status, :attributes, :events]
    keys = Map.keys(span)

    case keys -- supported_keys do
      [] -> :ok
      unsupported_keys -> raise "unsupported_keys = #{unsupported_keys}"
    end
  end

  defp get_current_span_id do
    case OpentelemetryTesting.get_span_ids() do
      {_trace_id, span_id} -> span_id
      :undefined -> :undefined
    end
  end

  defp link_to_pattern(link) do
    Map.take(link, [:trace_id, :span_id, :attributes])
  end

  defp set_status(status) do
    case status do
      :undefined -> :ok
      %{code: code, message: message} -> Tracer.set_status(code, message)
    end
  end

  defp create_new_link(attributes) do
    context = Tracer.current_span_ctx()
    random_attributes = pick_random_map_keys(attributes)
    OpenTelemetry.link(context, random_attributes)
  end

  defp pick_random_map_keys(:undefined) do
    ## covers :undefined span status case
    :undefined
  end

  defp pick_random_map_keys(map) do
    keys = Map.keys(map)
    drop_keys = pick_random_items(keys, length(keys), true)
    Map.drop(map, drop_keys)
  end

  defp pick_random_items([], _max_n, _allow_empty) do
    []
  end

  defp pick_random_items(_list, 0, _allow_empty) do
    []
  end

  defp pick_random_items(list, max_n, allow_empty) do
    length = length(list)

    n =
      case allow_empty do
        true -> :rand.uniform(min(length, max_n) + 1) - 1
        false -> :rand.uniform(min(length, max_n))
      end

    ## indexes must be unique and sorted in ascending order
    random_indexes =
      sequence(n) |> Enum.map(fn _ -> :rand.uniform(length) - 1 end) |> Enum.sort() |> Enum.uniq()

    Enum.map(random_indexes, &Enum.at(list, &1))
  end

  defp sequence(0), do: []
  defp sequence(n) when n > 0, do: 1..n
end
